### DISCUSSION TITLE
Trying to understand Urql's stream / exchange system in order to write a websocket client

### DISCUSSION THREAD

**Discussion by strblr** [urql-graphql/urql#1394] [Q&A]
Hello :) 

I'm trying to write a GraphQL websocket API with live query capacities, and then write an Urql exchange to hook to that API. On the client-side, my problems are twofold : 

1) Finding the best design strategy to dispatch updates from the server (should I use events ? Observables ? Subjects ?) ;
2) Understanding some core concepts of Urql like streams and teardown operations, in order to write the exchange without messing things up.

Heavily inspired by [this package](https://github.com/n1ru4l/graphql-live-query), the general idea for the client part would be to receive operations from some GraphQL client (Apollo or Urql), and give back a stream of results. One result stream per operation, one result per "live update" on that operation. The server would be stateful and exactly know which client is subscribed to what. One would then have an `invalidate` function (with roughly the same semantic than what you have on the graphcache), and that function would typically be called from mutations, or in database event listeners.

I want to do it both as an exercise and for I might need it in the future.

So, I spent the whole day looking at how your exchanges are written, reading the wonka doc, and some parts of the Urql doc. I've read the part on exchanges multiple times in the last days. And I have a hard time really getting to the bottom of it. It all seems logical and simple, but idk, something just doesn't want to click in my head when it comes to steams, observables, observers, forward, etc. It's as if there was something very simple to be understood right there in front of me, and I just can't reach it.

More specifically : 

- How does the client know which result in the result stream corresponds to which operation sent down the operation stream ?

- I understand that `teardown` ops are designed to signal to the exchange chain that there's no more interest in the result of an operation. But I don't understand why it's necessary. Wouldn't unsubscribing from the result stream and not sending anymore operations (from the kind you're no more interested in) down the pipe be enough ? On which exchange does `teardown` act and how so ?

- Back to my project, what would be the best way to be able to take one operation sent down the exchange pipe and return a stream of results only for that one operation ?

The package I mentioned earlier contains an Urql demo and the author [uses the `subscriptionExchange`](https://github.com/n1ru4l/graphql-live-query/blob/main/packages/todo-example/client-urql/src/createUrgqlClient.tsx), with `enableAllOperations` set to `true` ([here is the core implementation](https://github.com/n1ru4l/graphql-live-query/blob/main/packages/socket-io-graphql-client/src/createSocketIOGraphQLClient.ts) of the client part ; it uses async iterators and then some hack to make it push streams). Is this the only way to have the Urql client to subscribe to multiple results for one operation ?

- Another and last question might be : When there is a variable change in `useQuery` or `useSubscription`, is there no other choice than to create a whole new Observable in `forwardSubscription` ? Does `teardown` have any role to play here ?

Any other advice or insight from you would be very very welcome, as I'm very much a beginner when it comes to async streams.

I'm sorry if all this seems like very obvious things, but I fail to make it all click. So if you have some time, I would be pleased to hear from you. Thank you in advance.

**Comment by kitten** [comment]
Ok, these are a lot of questions, so for now I'll try to just address them one by one, but since this doesn't quite mention what you're looking to build I can't jump straight into a code sample 😅

> How does the client know which result in the result stream corresponds to which operation sent down the operation stream ?

There's an `Operation` on each result so the original keys can be compared. As seen here: https://formidable.com/open-source/urql/docs/api/core/#operationresult
So typically exchanges compare the original key of an operation to `result.operation.key` (the instance may change)

> But I don't understand why it's necessary. Wouldn't unsubscribing from the result stream and not sending anymore operations (from the kind you're no more interested in) down the pipe be enough

No, each exchange is a continuous stream of operations and returns a continuous stream of results. If you end this stream you'd end all results (which you can't because you're typically not the one subscribing but only transforming / mapping the streams)

> what would be the best way to be able to take one operation sent down the exchange pipe and return a stream of results only for that one operation ?

I'm not quite sure what this specific question is for because it depends on what you mean by "only for that one operation" 😅
However, if you look at the `fetchExchange` (or at the `subscriptionExchange` for that matter) you can see how they share operations and then create two separately filtered streams that are merged back; one is handled locally one forwards the operations to the next exchange in the pipeline: https://github.com/FormidableLabs/urql/blob/e5ea77e836b11be379327bee7abe74109d0adabc/packages/core/src/exchanges/fetch.ts#L16

This principle is also shown here: https://formidable.com/open-source/urql/docs/concepts/exchanges/#how-to-avoid-accidentally-dropping-operations

> it uses async iterators and then some hack to make it push streams). Is this the only way to have the Urql client to subscribe to multiple results for one operation ?

No, as you describe yourself, we use Observable-like streams. In fact, the subscription exchange takes care of a lot of this abstraction. It provides an API that you can provide with a function that receives the operation input and is supposed to return an `Observable`

> is there no other choice than to create a whole new Observable in forwardSubscription ? 

It depends. Each `Operation` is independent because in GraphQL each query + variables combination creates a new execution on an API (no matter the transport protocol)
So if you have some kind of grouping in mind then an exchange would have to do this grouping.
But in general, an exchange on its own gets more control over this than the `subscriptionExchange` provides.

**Reply by strblr** [reply]
A great thank you for the time you took to write this. It clarifies a lot for me, but I still have some grey zones and remaining questions if that's ok to you.

> No, each exchange is a continuous stream of operations and returns a continuous stream of results. If you end this stream you'd end all results

I don't know really why but I just assumed that new streams were created each time `useQuery` was used or something. So basically, if I got this right, the client instantiates two streams from the start, one to send operations, one to get results, and keeps these streams alive for all operations ?

- I still don't understand why `teardown` is necessary. Why is not sending any more ops of the kind you wanna end not enough ? Or has `teardown` some side-effects on some stateful exchanges ?

- Why does `subscriptionExchange` follow the pattern "one operation => stream of results" and not "stream of operations => stream of results" ?

- How can I notify Urql that there is an update for an existing query (the live query functionality that I want to implement), without Urql asking for it ?

I still haven't figured out how I should structure my project on the client-side, but the basic idea is to get operations, send them to the server, get results, and then getting further updates for these operations, in a push-stream manner. I'm gonna share some code samples when I'm 100% clear with how this should be done. I want to understand Urql first.

**Reply by kitten** [reply]
> Or has teardown some side-effects on some stateful exchanges ?

Exactly that. We need to be able to tell exchanges that nothing's interested and subscribed to an operation anymore. This could mean that the `fetchExchange` can for instance abort an ongoing API request. But it may also tell an exchange that an operation isn't relevant and onscreen anymore, so a refetch isn't required when the cache invalidates it for instance.

> Why does subscriptionExchange follow the pattern "one operation => stream of results" 

For the same reason that eventually each operation for a query ends up being one stream that wraps around a `fetch` request. Subscription abstractions commonly abstract sending an operation and receiving a result as per operation. They don't really have much of a choice since they're abstracting them as individual requests.

> How can I notify Urql that there is an update for an existing query (the live query functionality that I want to implement)

For a live query you'd just keep sending results until `urql` tears the operation down. If you use the `subscriptionExchange` you'd just send multiple updated results and you'd stop doing so and would cancel or unsubscribe from the query when the individual Observable is unsubscribed from.

**Reply by strblr** [reply]
Things really start to make sense for me. So, internally, `subscriptionExchange` will call `unsubscribe` on the operation's result Observable if a `teardown` is sent for that operation ? `subscriptionExchange` seems to have a lot more capabilities than just handling subscriptions. It's just another way of hooking into your Wonka streams, but without directly using them.

I might be able to start some coding. I'll keep you updated. Again, thank you.

**Reply by kitten** [reply]
> It's just another way of hooking into your Wonka streams, but without directly using them.

Yep, that's correct. With `enableAllOperations` it stops filtering and starts applying this pattern to all operations that represent requests, i.e. `mutation` and `query` too.

It has a large block that creates a new source for an operation: https://github.com/FormidableLabs/urql/blob/e5ea77e836b11be379327bee7abe74109d0adabc/packages/core/src/exchanges/subscription.ts#L64-L105
This is similar to the `fetchExchange`. It does something similar too. In the `subscriptionExchange` however it just converts the incoming operation to a source by using its API. So it converts it to the call that must return an observable and hooks it into Wonka streams.

You can see the typical stream creation there too. It calls `make` which is to create a `Source`. The inner function is called when the `Source` is subscribed to (so when the `Operation` starts in this case)
The logic to convert GraphQL results to `OperationResult` structures is abstracted too: https://github.com/FormidableLabs/urql/blob/e5ea77e836b11be379327bee7abe74109d0adabc/packages/core/src/exchanges/subscription.ts#L83-L84

It also converts the unsubscribe handler from Observable-likes to Wonka sources: https://github.com/FormidableLabs/urql/blob/e5ea77e836b11be379327bee7abe74109d0adabc/packages/core/src/exchanges/subscription.ts#L100-L103
And another small detail is that `teardown`s can also be sent actively. `client.reexecuteOperation` is a dispatch function for `Operation`s that's used by exchanges. It can be used to reexecute operations, i.e. to start a refetch, which will only work when the operation is still active. But here the `subscriptionExchange` sends an "active teardown" when the subscription ends: https://github.com/FormidableLabs/urql/blob/e5ea77e836b11be379327bee7abe74109d0adabc/packages/core/src/exchanges/subscription.ts#L86-L95

In the actual "Exchange block" of the code you can see that for each operation that matches a filter we then create the above source and merge it into the result stream (`mergeMap` merges each mapped source into a combined source): https://github.com/FormidableLabs/urql/blob/e5ea77e836b11be379327bee7abe74109d0adabc/packages/core/src/exchanges/subscription.ts#L117-L129

This `teardown$` stream is also a common pattern: https://github.com/FormidableLabs/urql/blob/e5ea77e836b11be379327bee7abe74109d0adabc/packages/core/src/exchanges/subscription.ts#L122-L125

When we say `takeUntil(teardown$)` we automatically limit the stream to stop when we receive a teardown operation for the original operation. This makes the stream stop when the `Client` isn't interested in it anymore.

**Reply by strblr** [reply]
Very clear, thank you !

**Reply by strblr** [reply]
Hello,

It may be unnecessary to open a new discussion for this, but I have one more question if I may.
Let's say the `operation` passed to `forwardSubscription` contains two queries :

```graphql
query MyQuery($id: ID!) {
  project(id: $id): Project!
  summary(id: $id): Summary!
}
```

What happens if I send a partial result to Urql's subscription observer ? For example only the result for `project`.

Ideally, I would want to be able to live-update only a subset of all the queries that can be contained in one operation. What would be the best approach here ? Should I instead force users to only send single-query operations ? Thanks !

**Reply by kitten** [reply]
I think you have a slight misunderstand here. "Single-query" operations is a bit of a misnomer since these are just fields, like any other field, on a type. You can't force users to split this up since that goes against what GraphQL is useful and used for. You'd liken them to subscriptions instead which are limited to one field per operation by the API.

> would want to be able to live-update only a subset of all the queries that can be contained in one operation

So, you can send partial results. But as I understand it what you're trying is to have a full result at the end? You'll have to merge these yourself if I understand the intention correctly.

**Reply by strblr** [reply]
Yes, queries are fields of the `Query` type, but I thought this was handled differently because it's a root type. On the server, I want to be able to invalidate specific queries, and then having the server rerun these queries and sending the results to the client even if that means sending partial results of a more complex operation. In the example I gave, I don't want to rerun both `project` and `summary` if only `project` is invalidated. So, for the very first result, the data would be complete. Only subsequent updates would be partial.

How to you handle invalidation ? Do you refetch the entire document even if only one query field was invalidated ?

**Reply by kitten** [reply]
This will fully depend on your transport protocol. You can of course create a protocol that allows you to patch the first result with the live updates as they come in. But at the end of the day the idea is that you'd recreate a result, as your schema guarantees here, since we're talking about live queries and not `@defer`. But to kind of tie this together: all the exchanges combined is all "business logic" that you get from `urql`, as in, it won't have any stitching of these results or other things like that since it's the responsibility of exchanges to accept operations and deliver results, which are sent straight to the bindings: https://formidable.com/open-source/urql/docs/concepts/philosophy/#declarative-queries

> How to you handle invalidation ? Do you refetch the entire document even if only one query field was invalidated ?

I'm not sure what you mean by invalidation. Like, I know you're referring to live queries but there's already an implementation you can use as reference. There's no standard on the transport protocol, but the principle is that whatever you transport, live queries intent is that you patch a first result you've sent continuously with new values.

**Reply by strblr** [reply]
Sorry my question wasn't precise enough : I was talking about `graphcache` invalidation. The `invalidate` method. I was wondering if invalidating a query field would trigger the entire operation to be refetched. But it seems the case, from what I understand. You don't slice operations into smaller chunks.

What live query implementation are you talking about ?

**Reply by kitten** [reply]
Yes and no, if we assume that an updater after a mutation invalidates a piece of data then that will retrigger the query operation. This operation will then go through Graphcache first again and only if then a cache miss occurs will it flow through (or fallthrough) to be executed against an API again.

It's not out of the question that another exchange could alter how query documents are sent though, especially with what we have planned for the `@populate` exchange. Such exchanges wouldn't carry through live query directives but it's something to keep in mind.

Anyway, if you're looking to implement live queries the interesting thing about them is that you wouldn't have to worry too much about Graphcache. All it'd have to do is to continuously update the first result it has sent and sent the patched results through too.
There may be some logic needed for reexecutions as you've said, but that comes down to some grouping and then depends on whether you'd want to restart the query or not against your API.
